tools/src/wyvern/target/corewyvernIL/type/RefinementType.java

```diff
@@ -190,32 +256,52 @@ public boolean equals(Object obj) {
         if (!(obj instanceof RefinementType)) {
             return false;
         }
+
         RefinementType other = (RefinementType) obj;
-        if (declTypes == null && other.declTypes == null) {
-            return base.equals(other.base) && typeParams.equals(other.typeParams);
+
+        if (!base.equals(other.base)) {
+            return false;
         }
+
         if (declTypes == null || other.declTypes == null) {
-            if (!base.equals(other.base)) {
-                return false;
-            }
-            return countRefinements() == other.countRefinements() && getParamList().equals(other.getParamList());
+            return getGenericArguments().equals(other.getGenericArguments());
+        } else {
+            return declTypes.equals(other.declTypes);
         }
-        return base.equals(other.base) && declTypes.equals(other.declTypes);
+
+        // TODO (@justinlubin) messes things up because of lazy evaluation?
```

Comment:
In which way does this piece of logic "messes things up"? Do test cases start to fail? Doesn't something else stop working? Please add more details about what goes wrong and how the commented out code is different from the new version of it.